### PR TITLE
automate releases

### PR DESCRIPTION
**Pull request by cammellos**
This automates the releases to github, when running `status-go-manual` step in jenkins.
the changes are:

- Release format on github has been changed to 
```
${version}/status-go-android.aar
${version}/status-go-ios.zip
${version}/status-go-desktop.zip
```

The release version is just the git tag, I would personally make it more explicit by providing a `VERSION` file (potentially with a script to update) containing the version to be uploaded (this will be the same as the `STATUS_GO_VERSION` in status-react), how do people feel about it?

https://ci.status.im/blue/organizations/jenkins/status-react%2Fpull%20requests/detail/PR-6579/2/pipeline/ is the status-react companion PR
and 
here is how it looks like on the releases page: 
https://github.com/status-im/status-go/releases/tag/v0.16.4-7-gd0f204dc

thanks @jakubgs for the help with credentials

<blockquote><div><strong><a href="https://ci.status.im/blue/organizations/jenkins/status-react%2Fpull%20requests/detail/PR-6579/2/pipeline/">Jenkins</a></strong></div></blockquote>
<blockquote><div> GitHub</div><div><strong><a href="https://github.com/status-im/status-go">status-im/status-go</a></strong></div><div>The Status module that consumes go-ethereum. Contribute to status-im/status-go development by creating an account on GitHub.</div></blockquote>

### CODE REVIEW

**Review by jakubgs — APPROVED** [review]
Nice!

File: Makefile
Comment (divan) [inline Makefile]: validate for empty `release_tag`, perhaps? (which will be the typical case if someone just tried to run `make prepare-release`)

File: Makefile
Comment (divan) [inline Makefile]: So do I understand correctly that anyone with repo write permission can just type `make deploy` and create a new release on github?
I wonder if we need to add more checks/questions here to prevent cases where people/scripts run `make deploy` just to see how it works (or by mistake). To me, publishing the release is an important thing, that need to be fool-proof and has to follow strict checklist. (I would  actually love to have multisig-like confirmation from other devs here, haha)

File: Makefile
Comment (adambabik) [inline Makefile]: Should we just out `clean-release` as a requirement?
```suggestion
prepare-release: clean-release
```

File: Makefile
Comment (adambabik) [reply Makefile]: I guess it's not needed because of `-f` flag.

File: Makefile
Comment (adambabik) [reply Makefile]: Yeah, multisig would be great. And I'd like to receive a push notification that there is a pending voting for me :D 

I would also think about some params validation and a message informing what's going to happen: `Deploying a release "$(release_tag)" for branch "$(release_branch)" from $(release_directory)`.

Also, `release_tag` seems unnecessary as it's not required by other commands but `/tmp/release-$(release_tag)` is assumed.

**Review by adambabik — CHANGES_REQUESTED** [review]
> The release version is just the git tag, I would personally make it more explicit by providing a VERSION file (potentially with a script to update) containing the version to be uploaded (this will be the same as the STATUS_GO_VERSION in status-react), how do people feel about it?

It's fine as long as we automate managing this file. This is tricky because we decide about a version when making `make deploy` by passing `release_tag` (I believe `VERSION` file and `release_tag` should be aligned). So, in theory, we should fix version in `VERSION` file, commit it and then make a release. After making a release, we should bump it again and commit to `develop`.

The alternative to this process might be relying solely on a value in `VERSION`. What I mean by that is when `make deploy` is run, we look at a value in `VERSION` and this is our tag. A git tag and GH release are created and then `VERSION` is bumped and committed to `develop`. What is not deterministic is the bump. We can default to patch and if it turns out a breaking change is introduced, a committer of that change would be responsible to bump minor.

File: Makefile
Comment (adambabik) [reply Makefile]: Oh and what do we need `release_branch` for?

Comment (cammellos) [reply Makefile]: We can add validation, `release_branch` is needed for github so it sets the right branch in the `archive` file (which we don't use currently, but it's created automatically by github)

File: Makefile
Comment (cammellos) [reply Makefile]: > The alternative to this process might be relying solely on a value in VERSION. What I mean by that is when make deploy is run, we look at a value in VERSION and this is our tag. A git tag and GH release are created and then VERSION is bumped and committed to develop. What is not deterministic is the bump. We can default to patch and if it turns out a breaking change is introduced, a committer of that change would be responsible to bump minor.

This would be my preferred workflow, I use that in a few projects, basically once we want to release, we change the `VERSION`, bumping by a minor/major depending on breaking changes, and then run make deploy, which solely looks at `VERSION`. We can run some safety check to make sure there is not already a release with that version (someone forgot to bump before releasing), allowing possibly to override (in case of broken releases etc).

I would not automate the bumping of `VERSION` (i.e. having the CI system do that), as I prefer this action to be explicit (we can of course provide a script to bump minor/major).

File: Makefile
Comment (jakubgs) [reply Makefile]: This is a very good point. Something like:
```Markdown
deploy:
	@read -p "Are you sure you want to create a new GitHub release? (y/n): " REPLY; \
	if [ $$REPLY = "y" ]; then \
	    github-release status-im/status-go $(release_tag) $(release_branch) "" "$(release_directory)"; \
	else \
	    echo "Aborting." && exit 1; \
	fi
```
BTW, I'd call this target `release` rather than `deploy`.

File: Makefile
Comment (adambabik) [reply Makefile]: > I would not automate the bumping of VERSION (i.e. having the CI system do that), as I prefer this action to be explicit (we can of course provide a script to bump minor/major).

The problem here is that after a release we will keep the old version and if someone creates alpha, beta etc., it will refer as `already.released.version-beta` while it should be `upcoming.release-beta`.

I agree with the rest. Let's create `VERSION` file then and simplify the workflow.

**Comment by cammellos** [comment]
@adambabik @divan incorporated the feedback, I have also updated the `README` with a description of the flow, let me know if there's anything you'd like to change.

File: Makefile
Comment (adambabik) [inline Makefile]: It probably should be uncommented before merging.

File: Makefile
Comment (adambabik) [inline Makefile]: After giving it a second though, maybe we can provide a default value for this?

```
release: RELEASE_BRANCH = develop
release: PRE_RELEASE = 0
release:
    @read -p "Are you sure you want to create a new GitHub $(shell if [[ "$$PRE_RELEASE" == "0" ]] ; then echo release; else echo pre-release ; fi) against $(RELEASE_BRANCH) branch? (y/n): " REPLY; \
    ...
```

File: _assets/ci/Jenkinsfile-manual
Comment (adambabik) [inline _assets/ci/Jenkinsfile-manual:82]: Is it gonna work if we have `if [ $$REPLY = "y" ];`? Also does it work without `echo` before `yes`?

File: Makefile
Comment (adambabik) [reply Makefile]: What's gonna happen if `RELEASE_TAG` already exists?

File: README.md
Comment (adambabik) [inline README.md]: `[...], GITHUB_TOKEN environment variable for example.`

File: Makefile
Comment (cammellos) [reply Makefile]: oh boy, sorry, forgot about that :)

File: Makefile
Comment (cammellos) [reply Makefile]: it will continue uploading the files, but it will stop if the files are already there, which seems like sensible behavior (we don't allow to override the assets, but in case they haven't uploaded at all, we allow to resume)

File: _assets/ci/Jenkinsfile-manual
Comment (cammellos) [reply _assets/ci/Jenkinsfile-manual:82]: seems like, tested on jenkins already (`yes` should output `y`, at least locally & on jenkins I believe) https://jenkins.status.im/job/status-go/job/status-go-manual/541/console

Comment (adambabik) [reply _assets/ci/Jenkinsfile-manual:82]: I was not aware of `yes` command O_o it keeps printing `y` in new lines.

File: Makefile
Comment (adambabik) [reply Makefile]: That sounds good!

File: _assets/ci/Jenkinsfile-manual
Comment (jakubgs) [reply _assets/ci/Jenkinsfile-manual:82]: Dude, you've missed out. Sometimes when I just don't know what to do anymore I just type `yes wtf` into the console.

File: Makefile
Comment (pedropombeiro) [inline Makefile]: Should we validate `$(release_branch)` to make sure it is not empty (e.g. typo in running `make release`)?

**Review by pedropombeiro — APPROVED** [review]

File: Makefile
Comment (cammellos) [reply Makefile]: it would throw an error as it's a required parameter, but we decided to default it to `develop`

**Comment by cammellos** [comment]
@adambabik updated, `PRE_RELEASE` defaults to `1`, and jenkins at the moment just creates `PRE_RELEASES`

**Review by adambabik — APPROVED** [review]
Automation FTW! 👏

**Pull request merged**
